packages/rollup-plugin/src/index.ts

```diff
@@ -0,0 +1,176 @@
+import { transpileFn } from '@typegpu/tgsl-tools';
+import type { AnyNode, VariableDeclarator } from 'acorn';
+import { walk } from 'estree-walker';
+import MagicString from 'magic-string';
+import type { Plugin } from 'rollup';
+
+const typegpuImportRegex = /import.*from\s*['"]typegpu.*['"]/g;
+const typegpuDynamicImportRegex = /import\s*\(\s*['"]\s*typegpu.*['"]/g;
+const typegpuRequireRegex = /require\s*\(\s*['"]\s*typegpu.*['"]\s*\)/g;
+
+type Context = {
+  /**
+   * How the `tgpu` object is used in code. Since it can be aliased, we
+   * need to catch that and act accordingly.
+   */
+  tgpuAliases: Set<string>;
+};
+
+type TgslFunctionDef = {
+  varDecl: VariableDeclarator;
+  implementation: AnyNode;
+};
+
+function embedJSON(jsValue: unknown) {
+  return JSON.stringify(jsValue)
+    .replace(/\u2028/g, '\\u2028')
+    .replace(/\u2029/g, '\\u2029');
+}
+
+function gatherTgpuAliases(ctx: Context, node: AnyNode) {
+  if (node.type === 'ImportDeclaration') {
+    if (
+      node.source.value === 'typegpu' ||
+      node.source.value === 'typegpu/experimental'
+    ) {
+      for (const spec of node.specifiers) {
+        if (spec.type === 'ImportDefaultSpecifier') {
+          // The default export of both 'typegpu' and 'typegpu/experimental' is the `tgpu` object.
+          ctx.tgpuAliases.add(spec.local.name);
+        } else if (
+          spec.type === 'ImportSpecifier' &&
+          spec.imported.type === 'Identifier' &&
+          spec.imported.name === 'tgpu'
+        ) {
+          // Aliasing 'tgpu' while importing, e.g. import { tgpu as t } from 'typegpu';
+          ctx.tgpuAliases.add(spec.local.name);
+        } else if (spec.type === 'ImportNamespaceSpecifier') {
+          // Importing everything, e.g. import * as t from 'typegpu';
+          ctx.tgpuAliases.add(`${spec.local.name}.tgpu`);
+        }
+      }
+    }
+  }
+}
+
+/**
+ * Checks if `node` is an alias for the 'tgpu' object, traditionally
+ * available via `import tgpu from 'typegpu'`.
+ */
+function isTgpu(ctx: Context, node: AnyNode): boolean {
+  let path = '';
+
+  let tail = node;
+  while (true) {
+    if (node.type === 'MemberExpression') {
```

Comment:
`node` is not modified inside this condition. if this condition is true and the one below it isn't, there will be an infinite loop